### PR TITLE
Adds Shockwaves to Explosions and Xeno Screeches 

### PR DESCRIPTION
**Pull request by GrrrKitten**
# About the pull request

Mostly ported from TGMC via code from Ivanmixo + Tiviplus

Tested and very functional, only activates on explosions with above 150 power meaning:

notable would shockwave:
-yajuta casters
-HE rockets
-nuke
-pod decompresion
-CAS minirockets
-All OB's
-most OT nades

notable wouldnt shockwave:
-AP rockets
-Mateba HE
-plastic explosives
-sentry explosions
-HEDP's
-HEFA
-other tiny explosions

Also adds shockwaves to:
-Queen Screech (very large)
-Predalien Screech (smaller)

# Explain why it's good for the game

Looks pretty cool, purely a VFX change via render plates. After extensively testing it for a few hours then going to play an actual round and seeing our OB's in their current state it was so sad. 

# Testing Photographs and Procedure

<summary>Screenshots & Videos</summary>

video had to be stupid compressed for github and also I cant crop it good cause photoshop said no more rendering today so have fun with this:
https://github.com/cmss13-devs/cmss13/assets/158611449/a83d26ea-3f7f-4053-8be4-d15ad707cdbc

Queen + Predalien Screech from 3 seperate POV's

https://github.com/cmss13-devs/cmss13/assets/158611449/137a496e-8b56-4039-ae84-cbaafa3de8d8

# Changelog

:cl:
add: Adds shockwave VFX to powerful explosions
add: Adds shockwave VFX to Queen + Predalien screech
/:cl:

### CODE REVIEW

File: code/game/objects/effects/temporary_visuals.dm [hunk 1]
```diff
@@ -96,3 +96,20 @@
...
+/obj/effect/temp_visual/shockwave
+	icon = 'icons/effects/light_overlays/shockwave.dmi'
+	icon_state = "shockwave"
+	plane = DISPLACEMENT_PLATE_RENDER_LAYER
+	pixel_x = -496
+	pixel_y = -496
+
+/obj/effect/temp_visual/shockwave/Initialize(mapload, radius)
+	. = ..()
+	deltimer(timerid)
```
Comment (fira) [inline code/game/objects/effects/temporary_visuals.dm:112]: If you're gonna do that you might want to consider just not using a temp_visual subtype at all, it's the only thing it does :P

**Review by fira — CHANGES_REQUESTED** [review]
also this is awesome

**Comment by Birdtalon** [comment]
Very cool.

**Comment by Biolock1** [comment]
This is awesome.

File: code/datums/autocells/explosion.dm [hunk 2]
```diff
@@ -282,6 +282,9 @@ as having entered the turf.
 	if(QDELETED(E))
 		return
 
+	if(power >= 150) //shockwave for anything over 150 power
+		new /obj/effect/temp_visual/shockwave(epicenter, power/60)
```
Comment (harryob) [inline code/datums/autocells/explosion.dm:286]:
```suggestion
		new /obj/effect/shockwave(epicenter, power/60)
```

File: code/game/objects/effects/temporary_visuals.dm [hunk 3]
```diff
@@ -96,3 +96,20 @@
...
+//------------------------------------------
+
+/obj/effect/temp_visual/shockwave
+	icon = 'icons/effects/light_overlays/shockwave.dmi'
+	icon_state = "shockwave"
+	plane = DISPLACEMENT_PLATE_RENDER_LAYER
+	pixel_x = -496
+	pixel_y = -496
+
+/obj/effect/temp_visual/shockwave/Initialize(mapload, radius)
```
Comment (harryob) [inline code/game/objects/effects/temporary_visuals.dm:110]:
```suggestion
/obj/effect/shockwave
	icon = 'icons/effects/light_overlays/shockwave.dmi'
	icon_state = "shockwave"
	plane = DISPLACEMENT_PLATE_RENDER_LAYER
	pixel_x = -496
	pixel_y = -496

/obj/effect/shockwave/Initialize(mapload, radius)
```

File: code/game/objects/effects/temporary_visuals.dm [hunk 4]
```diff
@@ -96,3 +96,20 @@
...
+	. = ..()
+	deltimer(timerid)
+	timerid = QDEL_IN_STOPPABLE(src, 0.5 * radius)
```
Comment (harryob) [inline code/game/objects/effects/temporary_visuals.dm:113]:
```suggestion
	QDEL_IN_STOPPABLE(src, 0.5 * radius)
```

File: code/__DEFINES/qdel.dm [hunk 5]
```diff
@@ -54,6 +54,7 @@
 #define QDESTROYING(X) (!X || X.gc_destroyed == GC_CURRENTLY_BEING_QDELETED)
 
 #define QDEL_IN(item, time) addtimer(CALLBACK(GLOBAL_PROC, GLOBAL_PROC_REF(qdel), (time) > GC_FILTER_QUEUE ? WEAKREF(item) : item), time, TIMER_STOPPABLE)
+#define QDEL_IN_STOPPABLE(item, time) addtimer(CALLBACK(GLOBAL_PROC, GLOBAL_PROC_REF(qdel), (time) > GC_FILTER_QUEUE ? WEAKREF(item) : item), time, TIMER_STOPPABLE)
```
Comment (harryob) [inline code/__DEFINES/qdel.dm:57]: redundant

**Review by harryob — DISMISSED** [review]

File: code/modules/mob/living/carbon/xenomorph/Abilities.dm [hunk 6]
```diff
@@ -142,7 +142,7 @@
 
 	playsound(xeno.loc, pick(xeno.screech_sound_effect_list), 75, 0, status = 0)
 	xeno.visible_message(SPAN_XENOHIGHDANGER("[xeno] emits an ear-splitting guttural roar!"))
-	xeno.create_shriekwave() //Adds the visual effect. Wom wom wom
+	xeno.create_shriekwave(xeno, 15) //Adds the visual effect. Wom wom wom, 15 shriekwaves
```
Comment (realforest2001) [inline code/modules/mob/living/carbon/xenomorph/Abilities.dm:145]: You pass the xeno here but don't actually use it in the proc

Comment (Drulikar) [reply code/modules/mob/living/carbon/xenomorph/Abilities.dm:145]: Refactor it out - you can already tell what xeno is shieking by looking at src

**Comment by realforest2001** [comment]
Off TM for now

File: code/game/objects/effects/temporary_visuals.dm [hunk 7]
```diff
@@ -96,3 +96,29 @@
...
+//------------------------------------------
+
+/obj/effect/shockwave
+	icon = 'icons/effects/light_overlays/shockwave.dmi'
+	icon_state = "shockwave"
+	plane = DISPLACEMENT_PLATE_RENDER_LAYER
+	pixel_x = -496
+	pixel_y = -496
+
+/obj/effect/shockwave/Initialize(mapload, radius, speed, easing_type, y_offset, x_offset)
```
Comment (Drulikar) [inline code/game/objects/effects/temporary_visuals.dm:110]:
```suggestion
/obj/effect/shockwave/Initialize(mapload, radius, speed, easing_type = LINEAR_EASING, y_offset, x_offset)
```

File: code/game/objects/effects/temporary_visuals.dm [hunk 8]
```diff
@@ -96,3 +96,29 @@
...
+	if(y_offset)
+		pixel_y += y_offset
+	if(x_offset)
+		pixel_x += x_offset
+	QDEL_IN(src, 0.5 * radius * speed)
+	transform = matrix().Scale(32 / 1024, 32 / 1024)
+	if(!easing_type)
+		animate(src, time = 0.5 * radius * speed, transform=matrix().Scale((32 / 1024) * radius * 1.5, (32 / 1024) * radius * 1.5))
+		return
+	animate(src, time = 0.5 * radius * speed, transform=matrix().Scale((32 / 1024) * radius * 1.5, (32 / 1024) * radius * 1.5), easing = easing_type)
```
Comment (Drulikar) [inline code/game/objects/effects/temporary_visuals.dm:123]:
```suggestion
	animate(src, time = 0.5 * radius * speed, transform=matrix().Scale((32 / 1024) * radius * 1.5, (32 / 1024) * radius * 1.5), easing = easing_type)
```

**Review by Drulikar — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/xenomorph/update_icons.dm [hunk 9]
```diff
@@ -196,31 +196,46 @@
...
+/mob/living/carbon/xenomorph/proc/create_shriekwave(shriekwaves_left)
+	var/offset_y = 8
+	if(mob_size == MOB_SIZE_XENO)
+		offset_y = 24
+	if(mob_size == MOB_SIZE_IMMOBILE)
+		offset_y = 28
+
+	//the shockwave center is updated eachtime shockwave is called and offset relative to the mob_size.
+	//due to the speed of the shockwaves, it isn't required to be tied to the exact mob movements
+	var/epicenter = src.loc ///center of the shockwave, set at the center of the tile that the mob is currently standing on
```
Comment (Drulikar) [inline code/modules/mob/living/carbon/xenomorph/update_icons.dm:208]:
```suggestion
	var/epicenter = loc //center of the shockwave, set at the center of the tile that the mob is currently standing on
```

**Comment by GrrrKitten** [comment]
tested and functional with the new commits

**Review by Drulikar — APPROVED** [review]

**Pull request merged**